render/tests/test-component.js

```diff
@@ -12,586 +12,1270 @@ o.spec("component", function() {
+
+				render(root, [])
+
+				o(called).equals(1)
+				o(root.childNodes.length).equals(0)
+			})
+			o("calls onbeforeremove", function() {
+				var called = 0
+				var component = {
+					onbeforeremove: function(vnode, done) {
+						called++
+
+						o(vnode.dom).equals(root.firstChild)
+						o(root.childNodes.length).equals(1)
+
+						done()
+					},
+					view: function() {
+						return {tag: "div", attrs: {id: "a"}, text: "b"}
+					}
+				}
+
+				render(root, [{tag: component}])
+
+				o(called).equals(0)
+
+				render(root, [])
+
+				o(called).equals(1)
+				o(root.childNodes.length).equals(0)
+			})
+			o("calls onbeforeremove when returning fragment", function() {
+				var called = 0
+				var component = {
+					onbeforeremove: function(vnode, done) {
+						called++
+
+						o(vnode.dom).equals(root.firstChild)
+						o(root.childNodes.length).equals(1)
+
+						done()
+					},
+					view: function() {
+						return [{tag: "div", attrs: {id: "a"}, text: "b"}]
+					}
+				}
+
+				render(root, [{tag: component}])
+
+				o(called).equals(0)
+
+				render(root, [])
+
+				o(called).equals(1)
+				o(root.childNodes.length).equals(0)
+			})
+			o("does not recycle when there's an onupdate", function() {
+				var component = {
+					onupdate: function() {},
+					view: function() {
+						return {tag: "div"}
+					}
+				}
+				var update = o.spy()
+				var vnode = {tag: component, key: 1}
+				var updated = {tag: component, key: 1}
+
+				render(root, [vnode])
+				render(root, [])
+				render(root, [updated])
+
+				o(vnode.dom).notEquals(updated.dom)
+			})
 		})
-		o("can remove when returning primitive", function() {
-			var component = {
-				view: function(vnode) {
-					return "a"
-				}
-			}
-			var div = {tag: "div", key: 2}
-			render(root, [{tag: component, key: 1}, div])
-			
-			render(root, [{tag: "div", key: 2}])
-			
-			o(root.childNodes.length).equals(1)
-			o(root.firstChild).equals(div.dom)
+		o.spec("state", function() {
+			o("deep copies state", function() {
+				var called = 0
+				var component = {
+					data: [{a: 1}],
+					oninit: init,
+					view: function() {
+						return ""
+					}
+				}
+
+				render(root, [{tag: component}])
+
+				function init(vnode) {
+					o(vnode.state.data).deepEquals([{a: 1}])
+				}
+			})
 		})
 	})
-	o.spec("lifecycle", function() {
-		o("calls oninit", function() {
-			var called = 0
-			var component = {
-				oninit: function(vnode) {
-					called++
-					
-					o(vnode.tag).equals(component)
-					o(vnode.dom).equals(undefined)
-					o(root.childNodes.length).equals(0)
-				},
-				view: function() {
-					return {tag: "div", attrs: {id: "a"}, text: "b"}
-				}
-			}
-			var node = {tag: component}
-			
-			render(root, [node])
-			
-			o(called).equals(1)
-			o(root.firstChild.nodeName).equals("DIV")
-			o(root.firstChild.attributes["id"].nodeValue).equals("a")
-			o(root.firstChild.firstChild.nodeValue).equals("b")
+
+	o.spec("class", function() {
```

Comment:
This is basically the same as the object tests, but adjusted for classes. The `create` helper is just sugar so I'm not repeating myself repeatedly creating the classes.